exotations/solvers/exotica_ilqr_solver/src/ilqr_solver.cpp

```diff
@@ -176,88 +189,119 @@ void ILQRSolver::Solve(Eigen::MatrixXd& solution)
+            const double& alpha = alpha_space(ai);
+            double rollout_cost = ForwardPass(alpha, ref_x, ref_u);
 
-            if (ai == 0 || (cost < current_cost && !std::isnan(cost)))
+            if (rollout_cost < cost_prev && !std::isnan(rollout_cost))
             {
-                current_cost = cost;
+                cost = rollout_cost;
                 new_U = prob_->get_U();
                 best_alpha = alpha;
+                break;
             }
         }
 
-        // finite checks
-        if (!new_U.allFinite() || !std::isfinite(current_cost))
+        // Finiteness checks
+        if (!new_U.allFinite() || !std::isfinite(cost))
         {
             prob_->termination_criterion = TerminationCriterion::Divergence;
-            WARNING_NAMED("ILQRSolver", "Diverged!");
+            WARNING_NAMED("ILQRSolver", "Diverged: Controls or cost are not finite.");
             return;
         }
 
+        time_taken_forward_pass = line_search_timer.GetDuration();
+
         if (debug_)
         {
-            HIGHLIGHT_NAMED("ILQRSolver", "Forward pass complete in " << line_search_timer.GetDuration() << " with cost: " << current_cost << " and alpha " << best_alpha);
-            HIGHLIGHT_NAMED("ILQRSolver", "Final state: " << prob_->get_X(T - 1).transpose());
+            HIGHLIGHT_NAMED("ILQRSolver", "Iteration " << iteration << std::setprecision(3) << ":\tBackward pass: " << time_taken_backward_pass << " s\tForward pass: " << time_taken_forward_pass << " s\tCost: " << cost << "\talpha: " << best_alpha);
         }
 
-        // copy solutions for next iteration
-        if (global_best_cost > current_cost)
+        //
+        // Stopping criteria checks
+        //
+
+        // Relative function tolerance
+        // (f_t-1 - f_t) <= functionTolerance * max(1, abs(f_t))
+        if ((cost_prev - cost) < parameters_.FunctionTolerance * std::max(1.0, std::abs(cost)))
         {
-            global_best_cost = current_cost;
-            last_best_iteration = iteration;
-            global_best_U = new_U;
-            best_ref_x_ = ref_x;
-            best_ref_u_ = ref_u;
+            // Function tolerance patience check
+            if (parameters_.FunctionTolerancePatience > 0)
+            {
+                if (iteration - last_best_iteration > parameters_.FunctionTolerancePatience)
+                {
+                    if (debug_) HIGHLIGHT_NAMED("ILQRSolver", "Early stopping criterion reached (" << cost << " < " << cost_prev << "). Time: " << planning_timer.GetDuration());
+                    prob_->termination_criterion = TerminationCriterion::FunctionTolerance;
+                    break;
+                }
+            }
+            else
+            {
+                if (debug_) HIGHLIGHT_NAMED("ILQRSolver", "Function tolerance reached (" << cost << " < " << cost_prev << "). Time: " << planning_timer.GetDuration());
+                prob_->termination_criterion = TerminationCriterion::FunctionTolerance;
```

Comment:
I'm not sure there shouldn't be a different stopping criteria -- `EarlyStopping`